infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java

```diff
@@ -258,4 +270,37 @@ private PersistentVolumeClaim replacePVCsWithCommon(
         .filter(subpath -> !isNullOrEmpty(subpath))
         .collect(Collectors.toSet());
   }
+
+  private void deleteCommonPVC(Workspace workspace) throws InfrastructureException {
+    factory.get(workspace).persistentVolumeClaims().delete(configuredPVCName);
+  }
+
+  private boolean userHasNoWorkspaces() {
+    boolean userHasNoWorkspaces = false;
+    String userId = getSessionUserIdOrUndefined();
+    if (!"undefined".equals(userId)) {
+      try {
+        Page<WorkspaceImpl> workspaces = workspaceManager.getWorkspaces(userId, false, 30, 0);
+        if (workspaces.isEmpty()) {
+          log.debug("User '{}' has no more workspaces left", userId);
+          userHasNoWorkspaces = true;
```

Comment:
Personally I prefer early return
```java
private boolean userHasNoWorkspaces() {
    String userId = getSessionUserIdOrUndefined();
    if (!"undefined".equals(userId)) {
      try {
        Page<WorkspaceImpl> workspaces = workspaceManager.getWorkspaces(userId, false, 1, 0);
        if (workspaces.isEmpty()) {
          log.debug("User '{}' has no more workspaces left", userId);
          return true;
        }
      } catch (ServerException e) {
        log.error(
            "Unable to get the number of workspaces for user '{}'. Cause: {}",
            userId,
            e.getMessage(),
            e);
      }
    }
    return false;
  }
```
Also, I don't think you need to fetch 30 workspaces, even one should be enough.

Reply:
good point regarding 30, but I preferred to use the default configuration.

Reply:
fixed

Reply:
> good point regarding 30, but I preferred to use the default configuration.

Since it's not a constant but just a hard-coded value, there is no guarantee that it will continue to be a default.

Reply:
right, but currently it is default on the service level, anyway changed to 1

Reply:
I think that is unnecessary logic. userid can be `che` in single-user mode or real in multi-user mode. It can't be "undefined"

Reply:
I suppose it may work in the wrong way when admin removes user and their workspaces.

> @sleshchenko @sparkoo @metlos wdyt. Maybe we should iterate over all workspace and check if that is the last workspace in the concrete k8s namespace. Of cause, if there are no workspaces at all it's safe to delete pvc.

it should fix the issue above

Reply:
Anonymus user will be "undefined" based on implementation.
Handling the case when the common PVC is used across multiple users ('workspaceNamespaceDefault' without placeholders is defined) - 8e02f78